Psr interfaces

kimpepper opened the pull request:
Fixes #199 

### Description
This PR removes the dependency on a specific HTTP client implementation and introduces PSR-7, PSR-17 and PSR-18 interfaces.

### Issues Resolved
It greatly simplifies the volume and complexity of the library codebase and removes the need to manage:
* client configuration: it can be now done directly on the implementing HTTP client, e.g. SSL, Basic Auth, Retries
* connection pools/host round-robin: this can be done with HTTP client middleware

Users of this library can choose from a number of synchronous and asynchronous HTTP clients including:
* Guzzle 7
* Symfony HTTP client

It greatly simplifies the volume and complexity of the library codebase. Specifically, it:
* removes a number of transient dependencies including `ezimuel/ringphp` which is not longer getting updates

Breaking changes
* We no longer json_decode the response body into an array.
* We leave that to the client to decide how they want to do that, e.g. deserialize into an object
* Clients can make use of `\Psr\Http\Message\StreamInterface` to decode the response in a memory efficient way

Tasks:
- [x] Replace Guzzle specific code with PSR interfaces
- [x] Remove unused code including Connection, ConnectionPool and CllientBuilder
- [x] Ensure test coverage

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock reviewed (changes requested):
I love this change, thank you! Let's iterate till we get this to green. I think it's a lot simpler than https://github.com/opensearch-project/opensearch-php/pull/228, what do you think @shyim?

- Sign DCO, `git commit -s --amend` for existing commits, force push.
- Add to CHANGELOG.
- Let's add an UPGRADING.md like https://github.com/opensearch-project/opensearch-py/blob/main/UPGRADING.md with details on migrating to this version.
- Should we keep a `OpenSearch\ClientBuilder` even with different parameters for easier migration?

dblock commented:
I noticed you tried to fix DCO?

```
"Kim Pepper [kim@previousnext.com.au](mailto:kim@previousnext.com.au)", but got "Kim Pepper [kim@pepper.id.au](mailto:kim@pepper.id.au)".
```

Something is not adding up between your settings and the commit signature.

kimpepper commented:
> I noticed you tried to fix DCO?
> 
> ```
> "Kim Pepper [kim@previousnext.com.au](mailto:kim@previousnext.com.au)", but got "Kim Pepper [kim@pepper.id.au](mailto:kim@pepper.id.au)".
> ```
> 
> Something is not adding up between your settings and the commit signature.

I think I fixed that now? Let me know if it's still not rigt.

I've updated the ClientBuilder and refactored the endpoints callable into it's own factory class. I think this makes it easier to understand, and also allows clients to swap out an implementation or use a dependency injection container if they want.

I updated the client and namespace templates also. I regenerated afterwards and it looks like there are a lot of new API methods added. Should I use something other than https://github.com/opensearch-project/opensearch-api-specification/releases/download/main-latest/opensearch-openapi.yaml ?

shyim commented:
I think setHandler are not working anymore right? So aws sig v4 needs to be adjusted?

kimpepper commented:
> I think setHandler are not working anymore right? So aws sig v4 needs to be adjusted?

Yes, I think we'd need to rewrite the guzzle middleware/handler without the ring dependency to make it easier for people who use AWS OpenSearch service to use signed requests.

dblock commented:
Let's work on merging https://github.com/opensearch-project/opensearch-php/pull/223 separately from this not to mix in the PSR changes? It needs some attention to tests.

dblock commented:
> > Something is not adding up between your settings and the commit signature.
> 
> I think I fixed that now? Let me know if it's still not rigt.

Yep looks good.
 
> I've updated the ClientBuilder and refactored the endpoints callable into it's own factory class. I think this makes it easier to understand, and also allows clients to swap out an implementation or use a dependency injection container if they want.

👍 
 
> I updated the client and namespace templates also. I regenerated afterwards and it looks like there are a lot of new API methods added. Should I use something other than https://github.com/opensearch-project/opensearch-api-specification/releases/download/main-latest/opensearch-openapi.yaml ?

Let's work on merging https://github.com/opensearch-project/opensearch-php/pull/223 separately from this not to mix in the PSR changes. It needs some attention to tests. If nobody picks it up here I can do it soon.

dblock commented:
Fixed CI in https://github.com/opensearch-project/opensearch-php/pull/234. Will merge and make a release with #223 next.

kimpepper commented:
Thinking about this more, returning HTTP Response might not be the best DX. 

We should introduce a `Response` object that contains the body of the HTTP Response and any meta-data. We can add a `toArray` method to return the deserialized JSON. We can have 2 sub-classes `ErrorResponse` and `OKResponse`. The error response can have `getErrorCode()` and `getErrorMessage()` methods.

The Promise would also return this too.

Thoughts?

dblock commented:
@kimpepper In other clients we return a strongly typed object that derives from a base class that also gives access to the underlying raw HTTP response from the network library "as is" along with some helper methods like status code or raw body. Note that OpenSearch returns more than JSON, too.

dblock commented:
I fixed what was broken in https://github.com/opensearch-project/opensearch-php/pull/223 and merged, rebase.

kimpepper commented:
> @kimpepper In other clients we return a strongly typed object that derives from a base class that also gives access to the underlying raw HTTP response from the network library "as is" along with some helper methods like status code or raw body. Note that OpenSearch returns more than JSON, too.

I think the strongly typed object would require a lot of work on the generator side and might be out of scope for this issue?

kimpepper commented:
From my understanding the response content types can be one of:

* plain text
* json
* nd-json
* xml

but I couldn't find the docs to back that up.

dblock commented:
> I think the strongly typed object would require a lot of work on the generator side and might be out of scope for this issue?

100%
 
> Not sure whether it's worth a simple wrapper around the PSR Response and HTTP Promise that gets returned?

I would if we think it makes it easier to extend it later in a common way across multiple transports (it feels like it would).

dblock commented:
> From my understanding the response content types can be one of:
> 
> * plain text
> * json
> * nd-json
> * xml
> 
> but I couldn't find the docs to back that up.

We cover the range in https://github.com/opensearch-project/opensearch-api-specification, see https://github.com/opensearch-project/opensearch-api-specification/blob/07e329e8d01fd0576de6a0a3c35412fd5a9163db/tools/src/tester/MimeTypes.ts#L10. 

OpenSearch returns plain text, JSON, nd-json, CBOR, SMILE, and YAML, most from [cat APIs](https://github.com/opensearch-project/opensearch-api-specification/blob/07e329e8d01fd0576de6a0a3c35412fd5a9163db/spec/namespaces/cat.yaml#L846).

kimpepper commented:
I created a separate PR #236 to deal with the bulk change to `static` return types.

kimpepper commented:
Split out #237 for the endpoint factory changes here.

dblock commented:
Thanks @kimpepper! Let's work through those.

stof commented on src/OpenSearch/EndpointFactory.php:
```diff
@@ -0,0 +1,57 @@
+     */
+    private function createEndpoint(string $class): EndpointInterface
+    {
+        $fullPath = '\\OpenSearch\\Endpoints\\' . $class;
```
This looks wrong to me if `$class` is already the `class-string`

stof commented on src/OpenSearch/Client.php:
```diff
@@ -65,214 +68,110 @@ class Client
-     * @var MlNamespace
-     */
-    protected $ml;
+    protected MonitoringNamespace $monitoring;
```
Turning protected properties into typed properties is technically a BC break for child classes if they redefine the properties.

Btw, if this is going to be released as a major version of the package, maybe private visibility should be used instead (removing the properties themselves from the API covered by semver as child classes won't have access to them anymore). After all, the child classes can read them using the public getter.

stof commented on src/OpenSearch/Namespaces/AbstractNamespace.php:
```diff
@@ -56,16 +68,23 @@ public function extractArgument(array &$params, string $arg)
+     *
+     * @throws \Psr\Http\Client\ClientExceptionInterface|\Exception
+     */
+    protected function performRequest(EndpointInterface $endpoint): Promise|ResponseInterface
```
This return type check actually changes the return type of all public methods on namespaces, which have not been updated

stof commented on src/OpenSearch/Client.php:
```diff
@@ -1714,16 +1613,15 @@ public function createPointInTime(array $params = [])
          *
          * @param array $params Associative array of parameters
-         * @return array
+         * @return \Http\Promise\Promise|\Psr\Http\Message\ResponseInterface
```
returning a ResponseInterface instead of parsing its body makes it impossible to document the structure of the response (defeating half of the goals of #217).
It also forces the caller to know how each endpoint is encoding its response (is it a JSON response or a plain text response ?), while the SDK could know that (this should be part of the API description)

stof replied:
This is also inconsistent with all other opensearch clients I looked at, which are all providing a proper typed response instead of a raw HTTP response.

kimpepper replied:
We don't generate a typed response unfortunately. As an intermediary step, we could return a new `OpenSearchResponseInterface` to allow us to create specific model responses in future.

We will need to support BC for the current `array` return param. We can add a new flag param `bool $returnArray = TRUE` that we trigger a deprecation if `true` and remove in 3.x. That will give users a chance to switch to expecting `OpenSearchResponseInterface`.

stof replied:
Returning an array shape can be described to static analyzers (this is what #217 is about).
Today, we don't actually return an array (contrary to what the phpdoc says) but a Ringphp ArrayAccess object (which is why I could not describe the array shape yet in the generated code as it would have required making Ringphp generic while it is unmaintained).

Returning an HTTP response leaves the user of the SDK responsible of knowing how an opensearch response looks like.

stof commented on src/OpenSearch/Client.php:
```diff
@@ -1977,34 +1868,57 @@ public function extractArgument(array &$params, string $arg)
+        $request = $this->transport->createRequest($method, $uri, $params, $body);
 
-        return $this->transport->resultOrFuture($promise, $options);
+        if ($this->isAsync()) {
```
instead of an instance-level flag (which separates the configuration from the usage, forcing all caller code to actually support both return types as they don't know how the client was configured), maybe this should be a argument of that method, allowing to use a conditional return type to be more precise (or even separate methods)

stof replied:
Looking at other clients in other languages, they often provide a *separate* async client (and corresponding async namespaces) for the async case, allowing to have proper return types for both cases instead of a union type.

stof commented on src/OpenSearch/Client.php:
```diff
@@ -1977,34 +1868,57 @@ public function extractArgument(array &$params, string $arg)
         );
-
-        return $this->transport->resultOrFuture($promise, $endpoint->getOptions());
+        if ($this->isAsync()) {
```
this `isAsync` flag changing the return type makes it impossible to properly define the return type of methods of the SDK, as they will change return type depending on the configuration of the client, which would be a pain for DX.

Also, as discussed in https://github.com/opensearch-project/opensearch-php/pull/219#discussion_r1715828607, it looks like the existing version of opensearch-php does not actually support using the async mode for those methods (unless I missed a way to enable it), so maybe this switch should be removed entirely here, to allow keeping a well-typed SDK.

kimpepper commented:
Thanks for your review @stof . I'm holding off any changes here until the dependent issue  #237 lands.

stof commented on composer.json:
```diff
@@ -21,22 +21,33 @@
     "ext-curl": "*",
-    "ezimuel/ringphp": "^1.2.2",
+    "ext-json": ">=1.3.7",
+    "php-http/async-client-implementation": "^1.0",
```
should we really require an async-client implementation when we might be using only the sync API ?

Maybe we should even go further than my previous comment and split the async and sync clients in separate packages, allowing them to have their own dependencies.

dblock replied:
It's a good idea, as long as it's not too much work. Generally, I am thinking introducing PSR interfaces is a key milestone that we need to hit so I'd want to make sure there's not crazy feature creep.

stof replied:
Well, reading my comment on https://github.com/opensearch-project/opensearch-php/pull/219#discussion_r1715828607 again, the existing client is sync-only, so the MVP is actually to keep this client sync-only for now. If there is a need for an async PHP client, it can be added later

dblock commented:
@kimpepper rebase let's see what this looks like?

kimpepper commented:
> @kimpepper rebase let's see what this looks like?

I think there is a fair bit to do if we are dropping async support in this PR.

kimpepper commented:
I did a basic rebase off `main` but this is still pretty rough. I'm not sure how we plan on handling BC.

Should we split off removing the async client?

kimpepper commented:
Thinking about this a bit further, we can keep the existing `Transport`, deprecate it, and introduce a new `PsrTransport`. 

Still need to figure out how to handle BC for the return types.

dblock commented:
This one is a major upgrade that is definitely worth it, sk I would introduce this as a breaking change. This is also the right time to change what these APIs return and really try to pack all the breakages in one major upgrade.

stof commented:
This PR seems to remove all support for sigv4. Shouldn't it provide a PSR client decorator implementing the sigv4 logic, as all other opensearch clients are shipping support for sigv4 to support the AWS managed opensearch service ?

dblock commented:
> This PR seems to remove all support for sigv4. Shouldn't it provide a PSR client decorator implementing the sigv4 logic, as all other opensearch clients are shipping support for sigv4 to support the AWS managed opensearch service ?

Yes, we definitely do not want to lose Sigv4 functionality. Breaking change is OK, but we need all the existing features.

kimpepper commented:
In the other issues we are trying to keep a BC layer for the 2.x branch. Are we saying this PR would go straight into a 3.x branch without any deprecations in 2.x?
 
If we try to get this in 2.x and provide BC this PR will:

* Drop async support (aparrently not working anyway according to [#219 (comment)](https://github.com/opensearch-project/opensearch-php/pull/219#discussion_r1715828607)
* Deprecate the `OpenSearch\ConnectionPool` and `OpenSearch\Connections` namespaces (can be provided by Psr client implementations)
* Deprecate `\OpenSearch\Transport`
* Create a new `\OpenSearch\PsrTransport`
* Deprecate `ClientBuilder`
* Create a new `\OpenSearch\PsrClientBuilder`
* Deprecate `\OpenSeach\Client`
* Create a new `\OpenSearch\PsrClient`
* `PsrClient` passes `PsrTransport $transport` param to `\OpenSearch\Namespaces\AbstractNamespace::__construct()`
* Handle BC and trigger deprecations in `\OpenSearch\Namespaces\AbstractNamespace::__construct()` for old `Transport`
* Change `\OpenSearch\Namespaces\AbstractNamespace::performRequest()` to _only_ return an array and drop it returning `FutureArrayInterface` or `callable` (as they aren't used)
* Change \OpenSearch\Handlers\SigV4Handler to work with Psr interfaces (follow up?)

This seems like quite a lot. Do you agree with the approach? Do you think we can split any of this out to sub-tasks or follow-ups?

dblock commented:
> In the other issues we are trying to keep a BC layer for the 2.x branch. Are we saying this PR would go straight into a 3.x branch without any deprecations in 2.x?

It should all go to main which will be the next major version of the client (3.0.0).

> Drop async support (aparrently not working anyway according to https://github.com/opensearch-project/opensearch-php/pull/219#discussion_r1715828607

If it's really not working we can take a delete in 2.x before this to trim the code.

Only breaking changes warrant a 3.0. If the sum of these can be done in a way of being deprecated -> added then we can continue doing it incrementally. I think you know best @kimpepper :)

kimpepper commented:
Yes, the intention is to provide deprecations in 2.x to allow users to upgrade to the Psr approach before 3.x is released.

kimpepper commented:
I've pushed new code to implement the concepts [above](https://github.com/opensearch-project/opensearch-php/pull/233#issuecomment-2484218103)

I haven't looked at tests yet.

stof commented:
@dblock async requests are working in the existing codebase for the method allowing to do a raw request (`Client::request`) but that's all. None of the generated SDK methods allow configuring the `async` option when calling `performRequest`, so the support for async requests in the Transport is only for `Client::request`. This PR is currently adding async in all methods though a global switch, making all return types blurry.

stof commented on src/OpenSearch/Aws/SigningRequestFactory.php:
```diff
@@ -0,0 +1,43 @@
+/**
+ * A decorator request factory that signs requests using the provided AWS credentials and signer.
+ */
+class SigningRequestFactory implements RequestFactoryInterface
```
This would be better implemented as a client decorator than as a request factory decorator. signing must happen *after* any transformation applied on the Request

stof commented on src/OpenSearch/LegacyTransportWrapper.php:
```diff
@@ -0,0 +1,34 @@
+        array $headers = [],
+    ): array|string|null {
+        $promise = $this->transport->performRequest($method, $uri, $params, $body);
+        return $this->transport->resultOrFuture($promise);
```
`resultOrFuture` will return a FutureArray of ringphp in case of sync mode (which is the case here). It does not actually return an array. This will fail with the native return type which is enforced by PHP. You need to extract the actual data (using `$futureArray->_value` to read it)

stof commented on src/OpenSearch/Namespaces/AbstractNamespace.php:
```diff
@@ -35,6 +39,8 @@ abstract class AbstractNamespace
 
     protected EndpointFactoryInterface $endpointFactory;
 
+    protected bool $isAsync = false;
```
should be removed

stof commented on src/OpenSearch/Endpoints/AbstractEndpoint.php:
```diff
@@ -22,26 +22,27 @@
      * @var array
      */
-    protected $params = [];
+    protected array $params = [];
```
add native types on protected properties is a BC break in case those properties are redefined in a child class

stof commented on src/OpenSearch/Endpoints/AbstractEndpoint.php:
```diff
@@ -83,9 +84,8 @@ abstract public function getMethod(): string;
      * Set the parameters for this endpoint
      *
      * @param mixed[] $params Array of parameters
-     * @return $this
```
I suggest keeping `@return $this`, which is a more precise contract

stof commented on src/OpenSearch/TransportFactory.php:
```diff
@@ -0,0 +1,131 @@
+
+    protected ?RequestFactoryInterface $requestFactory = null;
+
+    protected ?ClientInterface $httpClient = null;
```
I would suggest making those properties private, to simplify maintenance. Implementing a BC layer for protected properties is almost impossible in PHP (until we can target PHP 8.4+ which gives us property hooks for that).
As all properties have public setters and protected getters, this would not hurt extensibility for child classes.

stof commented on util/template/client-class:
```diff
@@ -67,13 +70,19 @@ class Client
+            $this->transport = $transport;
+            $this->httpTransport = new LegacyTransportWrapper($transport);
+        } else {
+            $this->httpTransport = $transport;
```
This technically changes the type of the public `$transport` property to `Transport|null` as it does not assign a transport.

stof commented on util/template/client-class:
```diff
@@ -132,35 +141,49 @@ class Client
     }
 
     /**
      * @return callable|array
```
this return type should be removed, as it is not true anymore.

stof commented on util/template/client-class:
```diff
@@ -132,35 +141,49 @@ class Client
+    /**
+     * Set the client to run in async mode.
+     */
+    public function setAsync(bool $isAsync): static
```
shouldn't those be removed ?

dblock commented:
> @dblock async requests are working in the existing codebase for the method allowing to do a raw request (`Client::request`) but that's all. None of the generated SDK methods allow configuring the `async` option when calling `performRequest`, so the support for async requests in the Transport is only for `Client::request`. This PR was adding async in all methods though a global switch, making all return types blurry.

Got it. I would remove it and see if anyone complains, but technically it's a breaking change. I will 🙈 if it gets removed.

dblock commented:
> @dblock async requests are working in the existing codebase for the method allowing to do a raw request (`Client::request`) but that's all. None of the generated SDK methods allow configuring the `async` option when calling `performRequest`, so the support for async requests in the Transport is only for `Client::request`. This PR was adding async in all methods though a global switch, making all return types blurry.

Btw, @stof any interest in ripping this out now before the Psr work on main?

kimpepper commented:
This PR seems to have stalled. Is there anything I can do to push it along?

dblock commented:
@kimpepper I think we're looking at you to get this to green? How can I help?

kimpepper commented:
Ah ok. I had an unanswered question for @stof . I'll look at the test fails.

kimpepper commented:
OK. Fixed the tests and phpstan errors. We will need to do another code generation for some of these changes before we commit. Should we do an API update in #241  first?

dblock commented:
Merged #241.

dblock commented:
To me this looks good, 👏 

Could I please get a CR from either @shyim and/or @stof, lets get this merged!

kimpepper commented:
Rebased on `main` and made a minor doc change.

dblock commented:
@kimpepper WDYT about adding something like https://github.com/opensearch-project/opensearch-py/blob/main/UPGRADING.md?

kimpepper commented:
Yep, I think that's a great idea.

kimpepper commented:
Added UPGRADING.md

stof commented on src/OpenSearch/Aws/SigningClientDecorator.php:
```diff
@@ -0,0 +1,34 @@
+        protected CredentialsInterface $credentials,
+        protected SignatureInterface $signer,
+    ) {
+        if (!class_exists(SignatureInterface::class)) {
```
This is broken. `SignatureInterfaceè is an interface, not a class, so this would always throw an exception when using it.

and the case of the SDK not being available cannot be handled inside the constructor: if there interface does not exists, there is no way to pass valid arguments to the constructor.

dblock commented:
Given the backwards incompatible changes I think this should bump us to 3.0, right?

kimpepper commented:
I added a lot of BC and deprecation triggers so this would go in 2.4.x and the BC layer removed in 3.0.x

dblock commented:
Cool, cool. @stof and @shyim let's get this merged? Help review?

kimpepper commented:
The only thing not mentioned is dropping async which wasn't working anyway.

dblock commented:
> The only thing not mentioned is dropping async which wasn't working anyway.

Add a line. Better safe than sorry.

kimpepper commented:
> Add a line. Better safe than sorry.

Added more details to the `CHANGELOG.md`

kimpepper commented:
I've been doing some manual testing of the new API with the [`drupal/search_api_opensearch`](https://www.drupal.org/project/search_api_opensearch) module to see how easy it is to use and so far so good.

kimpepper commented:
I needed to create my own factories for Guzzle and Symfony as we allow Drupal config to manage a few of the settings such as base URI and basic auth.

kimpepper commented:
Here's the issue https://www.drupal.org/project/search_api_opensearch/issues/3495196

kimpepper commented:
Happy new year! :tada:  Anything I can do to help move this forward?

dblock commented:
@stof @shyim any objections on merging this?

dblock commented:
@kimpepper are you in the public slack? https://opensearch.org/slack.html
join #clients I just mentioned this
